File: androidApp/src/main/kotlin/com/market/android/presentation/viewmodel/GroceryViewModel.kt
package com.market.android.presentation.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.market.shared.domain.model.Category
import com.market.shared.domain.model.GroceryItem
import com.market.shared.domain.usecase.GroceryUseCases
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.util.UUID

enum class FilterOption(val label: String) {
    ALL("Todos"),
    PENDING("Pendentes"),
    CHECKED("Feitos")
}

data class GroceryUiState(
    val allItems: List<GroceryItem> = emptyList(),
    val isLoading: Boolean = true,
    val filter: FilterOption = FilterOption.ALL,
    val searchQuery: String = ""
) {
    val filteredItems: List<GroceryItem>
        get() {
            val byFilter = when (filter) {
                FilterOption.ALL     -> allItems
                FilterOption.PENDING -> allItems.filter { !it.isChecked }
                FilterOption.CHECKED -> allItems.filter { it.isChecked }
            }
            return if (searchQuery.isBlank()) byFilter
            else byFilter.filter { it.name.contains(searchQuery, ignoreCase = true) }
        }

    val groupedItems: Map<Category, List<GroceryItem>>
        get() = filteredItems
            .sortedWith(compareBy({ it.isChecked }, { it.name }))
            .groupBy { it.category }
            .entries
            .sortedBy { it.key.ordinal }
            .associate { it.key to it.value }

    val checkedCount: Int get() = allItems.count { it.isChecked }
    val totalCount: Int get() = allItems.size
    val progress: Float get() = if (totalCount == 0) 0f else checkedCount / totalCount.toFloat()
}

class GroceryViewModel(private val useCases: GroceryUseCases) : ViewModel() {

    private val _filter = MutableStateFlow(FilterOption.ALL)
    private val _searchQuery = MutableStateFlow("")
    private val _isLoading = MutableStateFlow(true)

    val uiState = combine(
        useCases.getItems(),
        _filter,
        _searchQuery,
        _isLoading
    ) { items, filter, query, loading ->
        GroceryUiState(
            allItems = items,
            isLoading = loading,
            filter = filter,
            searchQuery = query
        )
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5_000),
        initialValue = GroceryUiState()
    )

    init {
        viewModelScope.launch {
            useCases.getItems().collect { _isLoading.value = false }
        }
    }

    fun addItem(
        name: String,
        quantity: Double,
        unit: String,
        category: Category,
        note: String
    ) = viewModelScope.launch {
        useCases.upsertItem(
            GroceryItem(
                id = UUID.randomUUID().toString(),
                name = name.trim(),
                quantity = quantity,
                unit = unit,
                category = category,
                isChecked = false,
                note = note.trim(),
                createdAt = System.currentTimeMillis()
            )
        )
    }

    fun updateItem(
        original: GroceryItem,
        name: String,
        quantity: Double,
        unit: String,
        category: Category,
        note: String
    ) = viewModelScope.launch {
        useCases.upsertItem(
            original.copy(
                name = name.trim(),
                quantity = quantity,
                unit = unit,
                category = category,
                note = note.trim()
            )
        )
    }

    fun toggleItem(item: GroceryItem) = viewModelScope.launch {
        useCases.upsertItem(item.copy(isChecked = !item.isChecked))
    }

    fun deleteItem(id: String) = viewModelScope.launch {
        useCases.deleteItem(id)
    }

    fun clearChecked() = viewModelScope.launch {
        useCases.clearChecked()
    }

    fun setFilter(filter: FilterOption) = _filter.update { filter }

    fun setSearchQuery(query: String) = _searchQuery.update { query }
}

File: iosApp/iosApp/Presentation/ViewModels/GroceryViewModel.swift
import Foundation
import Observation

// MARK: - Filter

enum FilterOption: String, CaseIterable, Identifiable {
    case all     = "Todos"
    case pending = "Pendentes"
    case checked = "Marcados"

    var id: String { rawValue }
}

// MARK: - ViewModel

@Observable
@MainActor
final class GroceryViewModel {

    // Raw list kept in sync with SwiftData
    private(set) var allItems: [GroceryItem] = []

    // UI controls
    var filter: FilterOption = .all
    var searchQuery: String  = ""

    // Confirmation dialogs
    var showClearConfirmDialog  = false
    var itemPendingDelete: String? = nil   // id of the item waiting for delete confirmation

    // Add / Edit sheet
    var showItemSheet   = false
    var editingItem: GroceryItem? = nil    // nil = new item

    private let store: GroceryStore

    init(store: GroceryStore = .shared) {
        self.store = store
        loadItems()
    }

    // MARK: Derived state

    var filteredItems: [GroceryItem] {
        var items = allItems

        switch filter {
        case .pending: items = items.filter { !$0.isChecked }
        case .checked: items = items.filter {  $0.isChecked }
        case .all:     break
        }

        if !searchQuery.isEmpty {
            items = items.filter { $0.name.localizedCaseInsensitiveContains(searchQuery) }
        }

        return items
    }

    var groupedItems: [(key: Category, value: [GroceryItem])] {
        let dict = Dictionary(grouping: filteredItems, by: \.category)
        return dict
            .map { (key: $0.key, value: $0.value) }
            .sorted { $0.key.rawValue < $1.key.rawValue }
    }

    var checkedCount: Int { allItems.filter(\.isChecked).count }
    var totalCount:   Int { allItems.count }

    var progress: Float {
        guard totalCount > 0 else { return 0 }
        return Float(checkedCount) / Float(totalCount)
    }

    // MARK: Actions

    func addItem(name: String, quantity: Double, unit: String, category: Category, note: String) {
        let item = GroceryItem(
            name:     name.trimmingCharacters(in: .whitespaces),
            quantity: quantity,
            unit:     unit,
            category: category,
            note:     note.trimmingCharacters(in: .whitespaces)
        )
        save(item)
    }

    func updateItem(_ original: GroceryItem, name: String, quantity: Double,
                    unit: String, category: Category, note: String) {
        let updated = GroceryItem(
            id:        original.id,
            name:      name.trimmingCharacters(in: .whitespaces),
            quantity:  quantity,
            unit:      unit,
            category:  category,
            isChecked: original.isChecked,
            note:      note.trimmingCharacters(in: .whitespaces),
            createdAt: original.createdAt
        )
        save(updated)
    }

    func toggleItem(_ item: GroceryItem) {
        var toggled = item
        toggled.isChecked.toggle()
        save(toggled)
    }

    func deleteItem(id: String) {
        try? store.delete(id: id)
        loadItems()
    }

    func confirmDelete(id: String) {
        itemPendingDelete = id
    }

    func clearChecked() {
        try? store.clearChecked()
        loadItems()
    }

    func openAddSheet() {
        editingItem  = nil
        showItemSheet = true
    }

    func openEditSheet(for item: GroceryItem) {
        editingItem  = item
        showItemSheet = true
    }

    // MARK: Private

    private func save(_ item: GroceryItem) {
        try? store.upsert(item)
        loadItems()
    }

    private func loadItems() {
        allItems = (try? store.fetchAll()) ?? []
    }
}
